File: 1319. Number of Operations to Make Network Connected/Solution.java
class Solution {
    public int makeConnected(int n, int[][] connections) {
        UFDS uf = new UFDS(n);
        for (int[] c : connections) uf.union(c[0], c[1]);
        return (uf.disjoint - 1 > uf.free) ? -1 : uf.disjoint - 1;
    }
}

class UFDS {
    int[] parent, rank;
    int disjoint, free;

    public UFDS(int n) {
        parent = new int[n];
        rank = new int[n];
        disjoint = n;
        free = 0;
        for (int i = 0; i < n; i++) parent[i] = i;
    }

    public int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    public boolean same(int x, int y) {
        return find(x) == find(y);
    }

    public void union(int x, int y) {
        if (same(x, y)) {
            free++;
            return;
        }
        int a = find(x), b = find(y);
        if (rank[a] < rank[b]) {
            parent[a] = b;
        } else {
            parent[b] = a;
            if (rank[a] == rank[b]) rank[a]++;
        }
        disjoint--;
    }
}

File: 1319. Number of Operations to Make Network Connected/Solution.py
class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        uf = UFDS(n)
        for x, y in connections: uf.union(x, y)
        return -1 if uf.disjoint - 1 > uf.free else uf.disjoint - 1

class UFDS:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.rank = [0 for i in range(n)]
        self.disjoint = n
        self.free = 0

    def find(self, x):
        if self.parent[x] != x: self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def same(self, x, y):
        return self.find(x) == self.find(y)

    def union(self, x, y):
        if self.same(x, y):
            self.free += 1
            return
        a, b = self.find(x), self.find(y)
        if self.rank[a] > self.rank[b]:
            self.parent[b] = a
        else:
            self.parent[a] = b
            if self.rank[a] == self.rank[b]: self.rank[b] += 1
        self.disjoint -= 1